core/include/jiminy/core/stepper/AbstractStepper.tpp

```diff
@@ -0,0 +1,50 @@
+
+#include "pinocchio/algorithm/joint-configuration.hpp"
+
+namespace jiminy
+{
+    template<typename F1>
+    AbstractStepper<F1>::AbstractStepper(F1 & f,
+                                         std::vector<Robot const *> robots):
+        f_(f);
+        robots_(robots)
+    {
+        // Empty
+    }
+
+
+    template<typename F1>
+    std::vector<vectorN_t> AbstractStepper<F1>::addToPosition(
+        std::vector<vectorN_t> const & positionIn,
+        std::vector<vectorN_t> const& velocityIn)
+    {
+        assert(positionIn.size() == robots_.size())
+        assert(velocityIn.size() == robots_.size())
+
+        std::vector<vectorN_t> positionOut;
```

Comment:
`positionOut` in input argument, and internal buffers for each knots to avoid memory allocation.

Reply:
Ah I see what you want to do.  In that case I think the best API is `add(q0, v, q1)` because  for Runge-Kutta because I have to do several times `qi = q0 + vi` with the same q0.
Won't there still be memory allocation though due to the fact that vectorN_t are not fixed sized ? How is `stdVect[i] = otherStdVect[i]``` handled, given that both Eigen::Vectors may have different sizes ? Should I enforce fixed size from the start.

Reply:
Also there is the question of doing a sum of more than one element: I could use successive addPosition() but this might not be very efficient... 
I think it would make sense to write it rather as a function with a variable number of arguments, which should be possible with fancy C++:

```
addPosition(qOut, q0, lambdai, q1, ... lambdaN qN): 
qOut = q0 + sum_i lambdai qi

```

Reply:
The problem is that I'm mixing types between lambda (scalar) and q (std::vector).
I could make use of std::pair, but this is likely to induce an unwanted overhead.

Reply:
> In that case I think the best API is add(q0, v, q1)

I agree.

> Won't there still be memory allocation though due to the fact that vectorN_t are not fixed sized ? How is `stdVect[i] = otherStdVect[i]``` handled, given that both Eigen::Vectors may have different sizes ? Should I enforce fixed size from the start.

You need one set of buffers for each system. So you need should define a `systemBuffers_t` struct to store them for each of them.

> I think it would make sense to write it rather as a function with a variable number of arguments, which should be possible with fancy C++:

Yes nice idea ! Just add a TODO and the method declaration and I could implement it using variadic template.

Reply:
Can variadic templates handle the fact that the signature is (type1, type2, type1, type2...) without creating a pair<type1, type2> ?

Reply:
Yes of course it is possible !

```
#include <type_traits>

struct A {};
struct B {};

template <typename X, typename Y, typename... Ts>
struct alternates : std::false_type {};

template <typename X, typename Y>
struct alternates<X, Y> : std::true_type {};

template <typename X, typename Y, typename... Ts>
struct alternates<X, Y, X, Ts...>
: std::conditional<alternates<Y, X, Ts...>::value, std::true_type, std::false_type>::type {};

struct C
{
    template<typename... Ts,
        typename = typename std::enable_if<
            sizeof...(Ts) % 2 == 0 &&
            sizeof...(Ts) >= 2 &&
            alternates<A, B, Ts...>::value
        >::type>
    C(Ts&&...);
};

int main() {
    C c1 { A(), B()};
    C c2 { A(), B(), A(), B(), A(), B() };
    // C c3 {};                     // I assumed you need at least 2
    // C c4 { A(), B(), A(), A() }; // A, A doesn't alternate
    // C c5 { B(), A(), B() };      // B, A, B not allowed
    // C c6 { A(), B(), A(), B() }; // A, B, A, B doesn't pair
}
```